### PR TITLE
TYP: Make temporary variables in pandas/__init__.py private

### PR DESCRIPTION
**Pull request by twoertwein**
Tiny step towards #39813 (`pyright --ignoreexternal --verifytypes pandas`) the relevant changes are in `pandas/__init__.py` and `pandas/_config/config.py` (private but `get_option`, ... uses the annotated wrapper).

I didn't plan to do the second commit (seemingly unrelated changes) but it seems that the first commit disturbed mypy's slumber (probably a few more cases that used the un-annotated wrapper?). I'm not very confident in the changes from the second commit (types for cython/c functions).

### CODE REVIEW

File: pandas/core/indexing.py [hunk 1]
```diff
@@ -935,7 +938,11 @@ def _getitem_lowerdim(self, tup: tuple):
             #  is equivalent.
             #  (see the other place where we call _handle_lowerdim_multi_index_axis0)
             with suppress(IndexingError):
-                return self._handle_lowerdim_multi_index_axis0(tup)
+                # error "_LocationIndexer" has no attribute
+                # "_handle_lowerdim_multi_index_axis0"
```
Comment (twoertwein) [inline pandas/core/indexing.py:942]: I was about to make _LocationIndexer an abstract class but a sub-class (I think _iLoc...) does not implement `_handle_lowerdim_multi_index_axis0` even though the class is being instantiated.

File: pandas/core/indexing.py [hunk 2]
```diff
@@ -935,7 +938,11 @@ def _getitem_lowerdim(self, tup: tuple):
...
+                return self._handle_lowerdim_multi_index_axis0(  # type: ignore[attr-defined]
```
Comment (twoertwein) [inline pandas/core/indexing.py:943]: any ideas on how to make flake8&mypy&black accept this line (and the same line later in this file)?

**Comment by jreback** [comment]
changes to `pandas/__init__.py` look fine (these could be separate PR if you want).

**Comment by twoertwein** [comment]
> changes to `pandas/__init__.py` look fine (these could be separate PR if you want).

I kept the simply changes in this PR and will later open a new PR for the remaining changes.

**Comment by twoertwein** [comment]
@jreback green

**Pull request merged**